# Jupyter Notebook for DS4400 Homework 1
## Advait Dharampal 

Problem 4: 
Generate at random 3 matrices of size $3 \times 3$ and fill each entry with a random integer chosen from -10 to 10. Use an existing package to compute the inverse of each matrix if it exists. Include in your report:
- Each of the 3 matrices
- The inverses computed with the package
- Compute the product of each matrix with its inverse to check that you obtain the identity matrix.

In [25]:
import numpy as np
import pandas as pd

In [22]:
# 3 Random Matrices
A = np.random.randint(-10, 11, size=(3,3))
B = np.random.randint(-10, 11, size=(3,3))
C = np.random.randint(-10, 11, size=(3,3))

print('Matrix A')
print(A)

print('\nMatrix B')
print(B)

print('\nMatrix C')
print(C)



Matrix A
[[ 7  1 10]
 [-6  6 -3]
 [ 0 -7  1]]

Matrix B
[[ 0 -8 -2]
 [ 8  2 -6]
 [ 8 -9 -1]]

Matrix C
[[ -2   6   9]
 [ -6   5  -3]
 [  2 -10   2]]


In [23]:
# Computing Inverse

# To avoid scientific notation, I am rounding to 4 decimal points
print('Inverse of A')
A_inv = np.linalg.inv(A)
print(np.round(A_inv, 4))

print('\nInverse of B')
B_inv = np.linalg.inv(B)
print(np.round(B_inv, 4))

print('\nInverse of C')
C_inv = np.linalg.inv(C)
print(np.round(C_inv, 4))

Inverse of A
[[-0.0467 -0.2212 -0.1963]
 [ 0.0187  0.0218 -0.1215]
 [ 0.1308  0.1526  0.1495]]

Inverse of B
[[-0.1129  0.0202  0.1048]
 [-0.0806  0.0323 -0.0323]
 [-0.1774 -0.129   0.129 ]]

Inverse of C
[[-0.038  -0.1939 -0.1198]
 [ 0.0114 -0.0418 -0.1141]
 [ 0.0951 -0.0152  0.0494]]


In [45]:
# Matrix Products
A_prod = np.round(A @ A_inv, 4)
print('Product of A and Inv:')
print(A_prod)

B_prod = np.round(B @ B_inv, 4)
print('\nProduct of B and Inv:')
print(B_prod)

C_prod = np.round(C @ C_inv, 4)
print('\nProduct of C and Inv:')
print(C_prod)


Product of A and Inv:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Product of B and Inv:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Product of C and Inv:
[[ 1.  0. -0.]
 [ 0.  1.  0.]
 [ 0. -0.  1.]]


Problem 5: The dataset for this assignment is available [here](https://drive.google.com/drive/folders/1FV7qofHD_olqygrIjASLQ7k7LTyVSEGG?usp=sharing). The prediction task is to predict the price of a house (column `price`) given the other features. Please ignore the columns `id` and `date`, as well as the categorical column `zipcode`. File `kc_house_data.csv` includes all the records in the dataset. You should use the entire dataset for the assignment. You can also find a Word document including the feature description in the same folder. In this problem, we will perform some exploratory data analysis using the house price dataset.

1. For each feature, write code to compute the average value, the min and max values, as well as its variance.
    
    Which features have the lowest and the highest average? Include the feature name and their average values for the features with the
    lowest and highest average.
    
    Which features have the lowest and the highest variance? Include the feature name and their variance values for the features with the
    lowest and highest variance.
    
2. Compute the correlation coefficient of each feature with the response. Include a table with the correlation coefficient of each feature with the response. 
Which features are positively correlated (i.e., have positive correlation coefficient) with the response?

Which feature has the highest positive correlation with the response?
3. Were you able to find any features with a negative correlation coefficient with the response?

In [41]:
# Cleaning the Data 
unclean_df = pd.read_csv('kc_house_data.csv')
data = unclean_df.drop(columns=['id', 'date', 'zipcode'])

# Getting the stats
stats = data.agg(['mean', 'min', 'max', 'var'])
print(stats)

lowest_mean_feature = stats.loc['mean'].idxmin()
lowest_mean = stats.loc['mean'].min()

print("\nLowest mean:", lowest_mean_feature)
print("Value:", lowest_mean)

highest_mean_feature = stats.loc['mean'].idxmax()
highest_mean = stats.loc['mean'].max()

print("\nHighest mean:", highest_mean_feature)
print("Value:", highest_mean)

lowest_var_feature = stats.loc['var'].idxmin()
lowest_var = stats.loc['var'].min()

print("\nLowest variance:", lowest_var_feature)
print("Value:", lowest_var)

highest_var_feature = stats.loc['var'].idxmax()
highest_var = stats.loc['var'].max()

print("Highest variance:", highest_var_feature)
print("Value:", highest_var)


             price   bedrooms  bathrooms    sqft_living      sqft_lot  \
mean  5.400881e+05   3.370842   2.114757    2079.899736  1.510697e+04   
min   7.500000e+04   0.000000   0.000000     290.000000  5.200000e+02   
max   7.700000e+06  33.000000   8.000000   13540.000000  1.651359e+06   
var   1.347824e+11   0.865015   0.593151  843533.681368  1.715659e+09   

        floors  waterfront      view  condition      grade     sqft_above  \
mean  1.494309    0.007542  0.234303   3.409430   7.656873    1788.390691   
min   1.000000    0.000000  0.000000   1.000000   1.000000     290.000000   
max   3.500000    1.000000  4.000000   5.000000  13.000000    9410.000000   
var   0.291588    0.007485  0.587243   0.423467   1.381703  685734.667269   

      sqft_basement     yr_built   yr_renovated        lat        long  \
mean     291.509045  1971.005136      84.402258  47.560053 -122.213896   
min        0.000000  1900.000000       0.000000  47.155900 -122.519000   
max     4820.000000  2015.

The lowest mean is longitude and its value is 0.007485225502686407, and the highest mean is price and its value is 540088.1417665294.

The lowest variance is waterfront and its value is 0.007485225502686407, and the highest variance is also price and its value is 134782378397.24687.

In [ ]:
# Getting the correlation coeeficants
price_corr = data.corr()['price'].sort_values(ascending=False)
print(price_corr)


price            1.000000
sqft_living      0.702035
grade            0.667434
sqft_above       0.605567
sqft_living15    0.585379
bathrooms        0.525138
view             0.397293
sqft_basement    0.323816
bedrooms         0.308350
lat              0.307003
waterfront       0.266369
floors           0.256794
yr_renovated     0.126434
sqft_lot         0.089661
sqft_lot15       0.082447
yr_built         0.054012
condition        0.036362
long             0.021626
Name: price, dtype: float64


Every feature is positevly correlated with Price, with the highest being sqft_living which has a coeeficent of 0.702035. I could not find any negative coeffiecnts with the response. 